plugin/trino-pulsar/src/main/java/io/trino/plugin/pulsar/PulsarConnectorCache.java

```diff
@@ -0,0 +1,213 @@
+                    (ns, offloader) -> {
+                        if (offloader != null && Objects.equals(offloader.getOffloadPolicies(), offloadPolicies)) {
+                            return offloader;
+                        }
+                        else {
+                            if (offloader != null) {
+                                offloader.close();
+                            }
+                            return initManagedLedgerOffloader(offloadPolicies, pulsarConnectorConfig);
+                        }
+                    });
+            managedLedgerConfig.setLedgerOffloader(ledgerOffloader);
+        }
+        return managedLedgerConfig;
+    }
+
+    private synchronized OrderedScheduler getOffloaderScheduler(OffloadPolicies offloadPolicies)
+    {
+        if (this.offloaderScheduler == null) {
+            this.offloaderScheduler = OrderedScheduler.newSchedulerBuilder()
+                    .numThreads(offloadPolicies.getManagedLedgerOffloadMaxThreads())
+                    .name("pulsar-offloader").build();
+        }
+        return this.offloaderScheduler;
+    }
+
+    private LedgerOffloader initManagedLedgerOffloader(OffloadPolicies offloadPolicies,
+                                                       PulsarConnectorConfig pulsarConnectorConfig)
+    {
+        try {
+            if (StringUtils.isNotBlank(offloadPolicies.getManagedLedgerOffloadDriver())) {
+                checkNotNull(offloadPolicies.getOffloadersDirectory(),
+                        "Offloader driver is configured to be '%s' but no offloaders directory is configured.",
+                        offloadPolicies.getManagedLedgerOffloadDriver());
+                Offloaders offloaders = offloadersCache.getOrLoadOffloaders(offloadPolicies.getOffloadersDirectory(),
+                        pulsarConnectorConfig.getNarExtractionDirectory());
+                LedgerOffloaderFactory offloaderFactory = offloaders.getOffloaderFactory(
+                        offloadPolicies.getManagedLedgerOffloadDriver());
+
+                try {
+                    return offloaderFactory.create(
+                            offloadPolicies,
+                            ImmutableMap.of(
+                                    LedgerOffloader.METADATA_SOFTWARE_VERSION_KEY.toLowerCase(), PulsarVersion.getVersion(),
+                                    LedgerOffloader.METADATA_SOFTWARE_GITSHA_KEY.toLowerCase(), PulsarVersion.getGitSha()),
+                            getOffloaderScheduler(offloadPolicies));
+                }
+                catch (IOException ioe) {
+                    log.error("Failed to create offloader: ", ioe);
+                    throw new RuntimeException(ioe.getMessage(), ioe.getCause());
+                }
+            }
+            else {
+                log.info("No ledger offloader configured, using NULL instance");
+                return NullLedgerOffloader.INSTANCE;
+            }
+        }
+        catch (Throwable t) {
```

Comment:
Please do not catch `Throwable` that may include OOM, or interrupted exceptions.